File: ruby/triangle/triangle.rb
# frozen_string_literal: true

class Triangle
  attr_reader :a, :b, :c

  def initialize(sides)
    @a, @b, @c = *sides
  end

  def equilateral?
    triangle_correct? && (a == b && a == c)
  end

  def isosceles?
    triangle_correct? && (a == b || a == c || b == c)
  end

  def scalene?
    triangle_correct? && (a != b && a != c && b != c)
  end

  private

  def triangle_correct?
    return false if a.nil? || b.nil? || c.nil?

    (a + b > c) && (b + c > a) && (a + c > b)
  end
end

File: python/triangle/triangle.py
def valid(fn):
    def wrapper(sides):
        a, b, c = sides
        valid_triangle = not (a == 0 or b == 0 or c == 0 or (a + b < c) or (a + c < b) or (b + c < a))
        return valid_triangle and fn(sides)
    return wrapper

@valid
def equilateral(sides):
    a, b, c = sides
    return a == b and b == c

@valid
def isosceles(sides):
    a, b, c = sides
    return a == b or b == c or c == a

@valid
def scalene(sides):
    a, b, c = sides
    return a != b and a != c and b != c
